lib/decanter/core.rb

```diff
@@ -1,212 +1,193 @@
+# frozen_string_literal: true
+
 module Decanter
   module Core
-    
     def self.included(base)
       base.extend(ClassMethods)
     end
 
     module ClassMethods
-      
-      def input(name, parsers=nil, **options)
-
-        _name = [name].flatten
+      # Declare an ordinary parameter transformation.
+      def input(name, parsers = nil, **options)
+        options[:type] = :input
+        options[:parsers] = parsers
+        handler(name, options)
+      end
 
-        if _name.length > 1 && parsers.blank?
-          raise ArgumentError.new("#{self.name} no parser specified for input with multiple values.")
-        end
+      # rubocop:disable Naming/PredicateName
 
-        handlers[_name] = {
-          key:     options.fetch(:key, _name.first),
-          name:    _name,
-          options: options,
-          parsers:  parsers,
-          type:    :input
-        }
+      # Declare a _has many_ association for a parameter.
+      def has_many(name, **options)
+        options[:type] = :has_many
+        options[:assoc] = name
+        handler(name, options)
       end
 
-      def has_many(assoc, **options)
-        handlers[assoc] = {
-          assoc:   assoc,
-          key:     options.fetch(:key, assoc),
-          name:    assoc,
-          options: options,
-          type:    :has_many
-        }
+      # Declare a _has one_ association for a parameter.
+      def has_one(name, **options)
+        options[:type] = :has_one
+        options[:assoc] = name
+        handler(name, options)
       end
+      # rubocop:enable Naming/PredicateName
+
+      # Add a parameter handler to the class.  Takes a name, and a set of
+      # options.  This is a generic method for any sort of hanlder, e.g.
```

Comment:
Fixed the typos :+1: 

Regards decanter against strong_arms, the big advantage of decanter is typing the parameters into integers, dates etc, for passing into subsequent services.  That's the bit that we were tying ourselves in knots with, rather than creating objects directly from them.  Hope that makes sense!